packages/core/saved-objects/core-saved-objects-base-server-internal/src/validation/validator.test.ts

```diff
@@ -127,39 +129,71 @@ describe('Saved Objects type validator', () => {
 
       data = createMockObject({ typeMigrationVersion: '3.5.0' });
       validator.validate(data, '4.5.0');
-      expect(getCalledVersion()).toEqual('4.3.0');
+      expect(getCalledVersion()).toEqual('3.0.0');
     });
 
     it('should use the correct schema for documents with typeMigrationVersion', () => {
-      let data = createMockObject({ typeMigrationVersion: '3.2.0' });
+      const data = createMockObject({ typeMigrationVersion: '3.2.0' });
+      validator.validate(data);
+      expect(getCalledVersion()).toEqual('3.0.0');
+    });
+
+    it('should use the correct schema for documents with typeMigrationVersion higher than default when not a valid virtual model version', () => {
+      const data = createMockObject({ typeMigrationVersion: '3.5.0' });
+      validator.validate(data);
+      expect(getCalledVersion()).toEqual('3.0.0');
+    });
+
+    it('should fall back to most recent schema for documents with migrationVersion', () => {
+      let data = createMockObject({
+        migrationVersion: {
+          [type]: '4.6.0',
+        },
+      });
       validator.validate(data);
       expect(getCalledVersion()).toEqual('3.0.0');
 
       jest.clearAllMocks();
 
-      data = createMockObject({ typeMigrationVersion: '3.5.0' });
+      data = createMockObject({
+        migrationVersion: {
+          [type]: '3.0.0',
+        },
+      });
       validator.validate(data);
-      expect(getCalledVersion()).toEqual('3.5.0');
+      expect(getCalledVersion()).toEqual('3.0.0');
     });
 
-    it('should use the correct schema for documents with migrationVersion', () => {
+    it('should use the correct schema for documents with migrationVersion higher than default when not a valid virtual model version', () => {
       let data = createMockObject({
         migrationVersion: {
           [type]: '4.6.0',
         },
       });
       validator.validate(data);
-      expect(getCalledVersion()).toEqual('4.3.0');
+      expect(getCalledVersion()).toEqual('3.0.0');
 
       jest.clearAllMocks();
 
       data = createMockObject({
         migrationVersion: {
-          [type]: '4.0.0',
+          [type]: '3.0.0',
         },
       });
       validator.validate(data);
-      expect(getCalledVersion()).toEqual('4.0.0');
+      expect(getCalledVersion()).toEqual('3.0.0');
+    });
+
+    it('should use the correct schema for documents with virtualModelVersion', () => {
+      const data = createMockObject({ typeMigrationVersion: '10.1.0' });
+      validator.validate(data);
+      expect(getCalledVersion()).toEqual('10.1.0');
+    });
+
+    it('should use the correct schema for documents with with virtualModelVersion higher than default when not a valid virtual model version', () => {
+      const data = createMockObject({ typeMigrationVersion: '11.1.4' });
+      validator.validate(data);
+      expect(getCalledVersion()).toEqual('3.0.0');
     });
```

Comment:
NIT: atm, `11.1.4` is an 'invalid' version for any document to have, given there's no reason to have a major superior to 10, so I wouldn't add tests about it tbh. (also, why does it use the `3.0.0` schema and not the `10.1.0` one?)

Reply:
> also, why does it use the 3.0.0 schema and not the 10.1.0 one?


The docVersion is `11.1.4`and the default version is 3.2.0,  and 11.1.4 isn't a valid virtualModeVersion. That means in:
`usedVersion = isVirtualModelVersion(docVersion) ? docVersion : this.defaultVersion;`
`previousVersionWithSchema` should be the most recent since `this.defaultVersion`, which is 3.2.0.
Since there isn't a 3.2.0, we end up validating against 3.0.0.